README
An R package to compute and plot the centre of mass of the tree itself. 

Estimating the coordinates of the centre of mass (barycentre) of a tree is crucial to judge its static stability. The centre of mass of a perfectly static tree lies inside its circumference at its base, that is the x and y coordinates of the barycentre lie inside the pi * r^2 surface where r is the radius of tree base.

The more distant the centre of mass from tree base the higher the constrains the tree poses on the soil through its roots. When concerns about tree stability are raised and  the tree needs to be consolidated a proper cabling system has to be put in place. Knowing in advance the direction the tree would fall in case of breakage at its base is necessary to properly engineer the cabling system.

Nevertheless, the estimate of barycentre position is useful to assess the effects of different pruning schemes as far as tree balance is concerned.

A simple plot of coefficient of slenderness of branches is also provided. Too thin/long branches and trees (i.e. whose SC > 50) are considered risky in the Visual Tree Assessment protocol. The plot aids in pointing out which branches exceed the SC50 rule and need to be pruned.

NEWS
1.2.3
Maintenance release: Vignette converted from latex to markdown format
1.2.2
Maintenance release: Updated to ensure compatibility with ggplot2 2.0.0 release
1.2.1
Maintenance release: Updated to ensure compatibility with recent ggplot2 releases
1.2
Maintenance release: Updated to provide compatibility with R 3.0
1.1
Removed switchBranchPruningStatus as it can be expressed in a simple base statement
Added feature to estimate the force on a ground plinth needed to stabilize a tree with a static steel cable
Added stonePine2FieldData dataset
Changed name of dataset fieldData to stonePine1FieldData
Changed name of dataset treeData to stonePine1TreeData
1.0.4
Updated to R 2.15.0
Updated to roxygen2
Slight change to treeVectors and new dependency from plyr
Correctly declared S3 methods: plot.vector, plot.SR and summary.CM
New S3 generic method plot.CM to plot CM on top of a plot generated by plot.vector
CM is now a named list instead of a vector
New S3 generic method print.CM prints just the cartesian coordinates of CM
Improved Makefile
1.0.3
Added a new allometry fitting equation yielding stone pine fresh weight (allometryABDC) 
Density dataset now provides 50% moisture content densities for a whole variety of conifers and dicots: Niklas, K. J. & Spatz, H.-C. Worldwide correlations of mechanical properties and green wood density American Journal of Botany, 2010, 97, 1587-1594
1.0.2
New feature: slenderness ratio for branches is computed and plotted. Tree SR is defined as L/D according to Mattheck, C. and Breloer, H. Banches SR is corrected to take into account branch deviation from the verticality.
Minor tweaks and speed improvements
Fixed bug that prevented data import
1.01 
fieldData column dir renamed to azimuth

R/CM.R
#' @title Computes the centre of mass of the tree
#'
#' @description The \eqn{x} coordinate of the centre of mass is defined as \eqn{\frac{\sum(m_ix_i)}{\sum(m_i)}} where \eqn{m_i} is the biomass of the \eqn{i^{th}} branch and \eqn{x_i} is the \eqn{x} coordinate of the \eqn{i^{th}} branch. \eqn{y} and \eqn{z} coordinates are similarly computed.
#' The centre of mass computation excludes branches to be pruned (ie: those whose \code{toBePruned} value is set to \code{TRUE}).
#'
#' @param object A data frame of class \code{vectors}
#' @return A vector holding \eqn{x}, \eqn{y}, \eqn{z} coordinates of the centre of mass
#' @export
centreOfMass <- function(object) {
  # sums the masses of tree branches and their x and y moments
  treeVectors <- object[!object$toBePruned, colnames(object) %in% c("Biomass", "mx", "my", "mz")]
  col.sums    <- apply(treeVectors, 2, sum)

  # cartesian coordinates of centre of mass of the tree
  M <- list(
    x = col.sums[["mx"]] / col.sums[["Biomass"]], 
    y = col.sums[["my"]] / col.sums[["Biomass"]], 
    z = col.sums[["mz"]] / col.sums[["Biomass"]]
  )
  class(M) <- c("CM", class(M))
  return(M)
}

#' @title Summary of Centre of Mass data
#'
#' @description Prints in a human-readable format the polar and cartesian coordinates of tree CM
#'
#' @param object An object of class \code{CM}
#' @param ...    Additional arguments, not used
#' @return       \code{NULL}
#' @method summary CM
#' @export
summary.CM <- function(object, ...) {
  polar <- toPolar(object$x, object$y)
  cat("Polar (angle/degrees, distance/m, height/m):", 
    polar[1], ",", 
    sprintf("%.2f", polar[2]), ",", 
    sprintf("%.2f", object["z"]), "\n"
  )
}

#' @title Simple print of Centre of Mass data
#'
#' @description Prints in a human-readable format the cartesian coordinates of tree CM
#'
#' @param x An object of class \code{CM}
#' @param ...    Additional arguments, not used
#' @return       \code{NULL}
#' @method print CM
#' @export
print.CM <- function(x, ...) {
    cat("Coordinates of the centre of mass:\n")
    cat("Cartesian (x/m, y/m, z/m):", 
      sprintf("%.2f", x$x), ",", 
      sprintf("%.2f", x$y), ",", 
      sprintf("%.2f", x$z), "\n"
      )
}

#' @title Plots tree CM
#'
#' @description Plots tree centre of mass as a layer on top of the \code{plot.vector}.
#'
#' CMs vector radii are proportional to CM magnitude. Tree CM is connected to tree base by an arrow showing the direction the tree would take in case of it falling down. \eqn{z} coordinate of tree CM is printed alongside its vector (if branch height has been recorded in the field).
#'
#' @param x      CM object
#' @param y      unused
#' @param ...    Arguments to be passed to plot.default
#' @return \code{NULL}
#' @method plot CM
#' @export
plot.CM <- function(x, y = NULL, ...) {
  chw <- par()$cxy[1] 
  cmText <- paste("CM (z=", sprintf("%.2f", x["z"]), ")")
  points(x$x, x$y, pch = 13, col = 2, cex = 3)
  text(
    x$x - chw, 
    x$y - chw, 
    labels = cmText, 
    adj = 0, 
    cex = 0.8,
    col = 2
    ) 
  arrows(0, 0, x$x, x$y, col = 2, lwd = 2) 
}

R/SC.R
#' @title Compute the slenderness ratio
#'
#' @description slenderness ratio is an important index of stability of trees and branches
#'
#' @note The coefficient takes into account branch angle: 
#' \eqn{SL_c=\frac{L}{D} \cdot (1 + cos \alpha)}, 
#' where \eqn{\alpha} is the branch angle (0 degrees = horizontal, 90 degrees vertical),
#' \eqn{L} is branch length in m, \eqn{D} is branch diameter in cm
#' Vertical branches have \eqn{SL = SL_c}
#'
#' @param x the data frame holding the measures needed to perform the computation
#' @param diameter The name of the data frame column holding diameter of the branch
#' @param length The name of the data frame column holding length of the branch
#' @param tilt The name of the data frame column holding tilt of the branch
#' @return slenderness ratio
#' @references Mattheck, C. and Breloer, H. \emph{The Body Language of Trees: A Handbook for Failure Analysis (Research for Amenity Trees)} 1995, HMSO (London)
branchSR <- function(x, diameter, length, tilt) {
  tiltRad <- as.double(x[tilt]) * pi / 180
  SR <- as.double(x[length]) / as.double(x[diameter]) * 100
  round(SR * ( 1 + cos(tiltRad)), digits = 0)
}

#' @title Computes slenderness ratio for tree branches
#'
#' @description slenderness ratio is an important index of stability of trees and branches
#'
#' @note The coefficient takes into account branch angle: 
#' \eqn{SL_c=\frac{L}{D} \cdot (1 + cos(a))}, 
#' where \eqn{a} is the branch angle (0 degrees = horizontal, 90 degrees vertical). 
#' Vertical branches have \eqn{SL = SL_c}
#'
#' @param treeObject an object of \code{treeData} class
#' @param vectorObject an object of \code{vectors} class
#' @return an object of class \code{SR}
#' @references Mattheck, C. and Breloer, H. \emph{The Body Language of Trees: A Handbook for Failure Analysis (Research for Amenity Trees)} 1995, HMSO (London)
#' @export
#' @seealso \code{\link{branchSR}}
treeSR <- function(treeObject, vectorObject) {
  SR <- treeObject$fieldData[, colnames(treeObject$fieldData) %in% c("dBase", "length", "tilt")]
  SR <- cbind(SR, vectorObject$Azimuth)
  SR <- cbind(SR, apply(SR, 1, branchSR, diameter = "dBase", length   = "length", tilt     = "tilt"))
  colnames(SR) <- c("diameter", "length", "tilt", "azimuth", "SR")
  class(SR) <- c("SR", class(SR))
  return(SR)
}


#' @title Plots slenderness ratio of branches
#'
#' @description Plots the branches as arrows whose length is proportional to their slenderness ratio.
#' A red circle holds ``safe'' branches (\eqn{SR_c \leq 70}).
#'
#' @note A circle is drawn to encompass 
#' the 70- values for slenderness ratio. Branches with 70+ values for the slenderness
#' ratio are considered dangerous. Please note that Mattheck coefficient is corrected to account 
#' for branch tilt (the more it deviates from the verticality the higher its coefficient) 
#'
#' @param x      SR object
#' @param y      unused
#' @param safeSR SR threshold, risky branches are red-coloured
#' @param ...    Arguments to be passed to plot.default
#' @return \code{NULL}
#' @method plot SR
#' @seealso \code{\link{treeSR}}
#' @export
plot.SR <- function(x, y = NULL, safeSR = 70, ...) {
  
  Circle <- function(t, a) {
    a * cos(t) + 1i * a * sin(t)
  }
  
  xy <- toCartesianXY(x$azimuth, x$SR)
  xyL <- length(xy)
  xyCoord <- data.frame(cbind(xy[1:(xyL/2)], xy[((xyL/2)+1):xyL]))
  colnames(xyCoord) <- c("x", "y")
  rownames(xyCoord) <- rownames(x)
  
  plot(xyCoord, type="n", asp = 1, ...)
  chw <- par()$cxy[1] 
  
  xyCoord <- cbind(xyCoord, x$SR)
  colnames(xyCoord) <- c("x", "y", "SR")
  safe   <- xyCoord[xyCoord$SR <= safeSR, colnames(xyCoord) %in% c("x", "y")]
  unsafe <- xyCoord[xyCoord$SR > safeSR, colnames(xyCoord) %in% c("x", "y")]
  
  if (nrow(safe) > 0) {
    arrows(0, 0, safe$x, safe$y, col = "green")
    text(safe$x - chw, safe$y - chw, labels = row.names(safe), adj = 0, cex = 0.8, col = "green") 
  }
  if (nrow(unsafe) > 0) {
    arrows(0, 0, unsafe$x, unsafe$y, col = "red")
    text(unsafe$x - chw, unsafe$y - chw, labels = row.names(unsafe), adj = 0, cex = 0.8, col = "red") 
  }
  
  t <- seq(0, 2 * pi, by=0.01)
  center <- 0 + 0i
  lines(center + Circle(t, safeSR), col = "red", lwd = 2)
}



R/math.R
#' @title Computes the x,y cartesian coordinates
#'
#' @description Computes the \eqn{x} and \eqn{y} cartesian coordinates from a set of polar coordinates
#'
#' @param angle The angle in degrees (measured clockwise from the North or  
#' any other relevant bearing system defined in the field)
#' @param distance The distance
#' @return A vector holding the \eqn{x} and \eqn{y} coordinats expressed in the same unit as the distance argument
#'
#' @note The function assumes the angle is measured clockwise whereas
#' trigonometric functions require a conventional counterclockwise 
#' measured angle. Thus the function computes \eqn{x} coordinate as the sine of 
#' the angle, and the \eqn{y} coordinate as the cosine of 
#' the angle, enabling a correct representation of them on a cartesian plot.
#' @export
#' @author Marco Bascietto \email{marco.bascietto@@ibaf.cnr.it}
toCartesianXY <- function(angle, distance) {
  angleRad <- angle * pi / 180
  c(sin(angleRad) * distance, cos(angleRad) * distance)
}


#' @title Converts cartesian (x, y) into polar (angle, distance) coordinates
#'
#' @description Converts cartesian coordinates (\eqn{x}, \eqn{y}) into polar (angle, distance) ones, assuming (0, 0) as origin of axes and, incidentally, the position of tree base
#'
#' @param x Abscissa coordinate
#' @param y Ordinate coordinate
#' @return A vector holding angle in degrees and distance in the same unit as \eqn{x} and \eqn{y}
#' @export
#' @author Marco Bascietto \email{marco.bascietto@@ibaf.cnr.it}
toPolar <- function(x, y) {
  d <- sqrt(x^2 + y^2)
  a <- (atan2(x, y) * 180 / pi) %% 360
  c(as.integer(a), d)
}


R/moment.R
#' @title Returns the coordinates of centre of mass of branches and logs
#'
#' @description Computes the cartesian coordinates of centre of mass of branches and 
#' logs along with their \eqn{x}, \eqn{y}, \eqn{z} moments
#'
#' The \eqn{x} and \eqn{y} coordinates are computed from the polar coordinates (angle and distance, 
#' defined as the length of its projection on ground), measured in the field. 
#' The \eqn{z} coordinate is computed by adding the height of branch insertion on the stem 
#' (measured in the field) to the height of the branch (calculated through its 
#' mean tilt, in case it was measured in the filed).
#' The \eqn{x}, \eqn{y}, \eqn{z} coordinates are corrected to take into account where the actual
#' centre of mass lies on the branches themselves by multiplying them by branchesCM,
#' a real number from 0.01 (CM at branch base) to 1.00 (CM at branch tip). 
#' As a rule of thumb, average live branches, with an average amount of foliage, 
#' have CM approx. \eqn{2/3} of their length, ie. branchesCM = 0.66.
#' \eqn{x}, \eqn{y}, \eqn{z} moments are computed by multiplying the corresponding cartesian coordinates by 
#' branch or log mass, e.g. \eqn{m_x = F \cdot l_x}, where \eqn{F} is branch or log mass, \eqn{l_x} is the \eqn{x} component of the lever arm (e.g. the \eqn{x} component of the branch or log projection on the ground).
#'
#' @param azimuth Branch compass heading
#' @param dBase unused argument
#' @param dTip unused argument
#' @param length Branch length
#' @param tipD unused argument
#' @param height Height of branch insertion on the stem or the height of log lower section
#' @param tilt Inclination of the branch or log in degrees
#' @param toBePruned unused argument
#' @param biomass Mass of the branch or log
#' @param branchesCM a real number varying from 0.01 to 1 proportional to the centre of
#' mass position along the branch (0.01 branch base, 1 branch tip)
#' @return a vector holding 5 reals:
#' \itemize{
#'  \item{the \eqn{x} coordinate of branch CM}
#'  \item{the \eqn{y} coordinate of branch CM}
#'  \item{the \eqn{x} moment of the branch}
#'  \item{the \eqn{y} moment of the branch}
#'  \item{the \eqn{z} moment of the branch}}
#' @note BranchCM is assumed to have same value in branches and logs. This is not the case in the real world. As a measure of safety one should use the highest value possible, eg branchesCM = 1.
#'
#' @note \eqn{z} coordinate of CM is not returned because it would be useless in a 2D plot. It is computed using \eqn{mz}, which is, as a matter of facts, returned
getCoordinatesAndMoment <- function (azimuth, dBase, dTip, length, tipD, height, tilt, toBePruned, biomass, branchesCM) {
  # height (h) to be added to branch height (z), as a function of the 
  # angle of its tilt (0° = horiz., 90° = vert.), its distance (length of its 
  # projection on the ground, 
  # from tree base to branch tip), and the estimated position of the centre of mass
  h  <- tipD * sin(tilt * pi / 180) * branchesCM
  ## computes cartesian coordinates of centre of mass of branches and their moments (mx, my, mz).
  ## When branchesCM = 1, x and y are coordinates of branch tip
  xy  <- toCartesianXY(azimuth, (tipD * branchesCM))
  z   <- height + h
  
  mx  <- biomass * xy[1]
  my  <- biomass * xy[2]
  mz  <- biomass * z
  c(xy, mx, my, mz)
}


R/vectors.R
#' @title Computes cartesian coordinates and moments of branches and logs 
#'
#' @description A data frame is populated with branch and log masses, along with \eqn{x}, \eqn{y} cartesian coordinates and \eqn{x}, \eqn{y}, and \eqn{z} moments.
#' \eqn{z} coordinates and moments are calculated only if branches height from the ground (and tilt) have been measured in the field.
#'
#' @param object an object of \code{treeData} class
#' @return an object of class \code{vectors} 
#' @seealso \code{\link{getCoordinatesAndMoment}}
#' @export
treeVectors <- function(object) {
  ## vectors data frame is populated
  vectors <- object$fieldData[, colnames(object$fieldData) %in% c("azimuth", "tipD", "biomass", "height", "tilt", "toBePruned")]

  ## computes cartesian coordinates of branch tip and its x, y and z moments are added to vectors slot
  vectors <- mdply(vectors, getCoordinatesAndMoment, branchesCM = object$branchesCM)
  colnames(vectors) <- c("Azimuth", "Distance", "Height", "Tilt", "toBePruned", "Biomass", "x", "y", "mx", "my", "mz")
  class(vectors) <- c("vector", class(vectors))
  return(vectors)
}

#' @title Plots branches and logs
#'
#' @description Plots branches and logs
#'
#' The 2d plot represents branches and logs as vectors pointing inwards.
#' Branches to be pruned are not shown on graph.
#'
#' @param x      vectors object
#' @param y      unused
#' @param txtcol Colour of text labels, defaults to "grey80"
#' @param ...    Arguments to be passed to plot.default
#' @return \code{NULL}
#' @method plot vector
#' @export
plot.vector <- function(x, y = NULL, txtcol = "grey80", ...) {
  treeVectors <- x[!x$toBePruned,]

  ## plots branch masses
  ## size of points is proportional to branch biomass
  maxPointSize <- 10
  pointSize <- maxPointSize * as.numeric(as.character(treeVectors$Biomass)) / max(treeVectors$Biomass)

  plot.default(treeVectors[c("x", "y")], cex = pointSize, pch = 13, ...)
  abline(h = 0, v = 0, col = "gray70")
  
  # print vector labels
  chw <- par()$cxy[1] 
  text(treeVectors[c("x", "y")] - chw, labels = row.names(treeVectors), adj = 0, cex = 0.8, col = txtcol) 
}


R/treecm.R
###############################################################################
#
# treecm: An R package to assess tree centre of mass
# author: Marco Bascietto <marco.bascietto@ibaf.cnr.it>
#
# This is released under a GPL license.
#
# Documentation was created using roxygen:
# package.skeleton(name="treecm", code_files="treecm.R")
# library(roxygen2); roxygenize('treecm', 'treecm', overwrite=TRUE, unlink.target=FALSE, copy.package=FALSE)
###############################################################################

#' @title Assessment of the position of the centre of mass of trees
#'
#' @description The centre of mass is a crucial data for arborists in order to consolidate a tree using steel or dynamic cables. Given field-recorded data on branchiness of a tree, the package:
#' \itemize{
#'   \item{computes and plots the centre of mass of the tree itself}
#'   \item{simulates the shift in CM position as branches are pruned}
#'   \item{computes branches slenderness coefficient in order to aid the arborist identify potentially dangerous branches}
#'   \item{computes the force acting on a ground plinth and its best position relating to the tree centre of mass, should the tree need to be stabilized by a steel cable}
#' }
#' The tree stem is ideally sectioned in logs. The weight of tree components is assessed based on
#' \itemize{
#'   \item the sum of volume of stem logs
#'   \item the sum of branches biomass
#' }
#' Field measures to be taken on logs and branches are described in \code{\link{importFieldData}} and are to be recorded on the tree itself, possibly using tree-climbing tecniques.
#' In order to help the arborist in the pruning selection process a simple plot of branch coefficient of slenderness is implemented.
#' @note \bold{Branch biomass} is computed by allometric equations relating its fresh weight (wood + leaves) to its diameter at point of insertion on the stem. \bold{Log biomass} is computed by converting its volume to weight using wood fresh density. Volume is computed using Smalian's formula (see \code{\link{logBiomass}} description).
#' A sample \code{.CSV} file is provided to guide through data filling in the field
#' @seealso \code{\link{logBiomass}}
#' \code{\link{importFieldData}}
#' @name treecm-package
#' @aliases treecm
#' @docType package
#' @import plyr
#' @author Marco Bascietto \email{marco.bascietto@@cnr.it}
#' @keywords package
#' @examples
#' data(stonePine1TreeData)
#' vectors  <- treeVectors(stonePine1TreeData)
#' CM       <- centreOfMass(vectors)
#' plot(vectors, 
#'    main = "Centre Of Mass", 
#'    col = "grey30", 
#'    txtcol = "grey30")
#' plot(CM)
#' summary(CM)
#' @references Source code is hosted at GitHub (\url{http://mbask.github.com/treecm/})
NULL




#' Green wood density data for a few tree species
#' 
#' Wood density is used to convert wood volume measures in the field to their
#' fresh weights. It is measured in \eqn{\frac{kg}{m^3}}.
#' 
#' Density is measured at humidity level 50%, very close to mean living tree
#' wood humidity. The dataset is provided as a reference only, please be
#' cautioned about using these values on your samples.
#' 
#' @name Dst
#' @docType data
#' @format A data frame with 170 observations on the following 3 variables. \code{
#' data.frame:  170 obs. of  3 variables:
#'   $ species: chr  "Abies alba" "Abies alba" "Abies balsama" "Abies grandis" ...
#'   $ group  : Factor w/ 2 levels "conifer","dicot": 1 1 1 1 1 1 1 1 1 1 ...
#'   $ density: int  545 577 529 449 465 673 689 497 673 577 ...}
#' @source Niklas, K. J. and Spatz, H.-C. (2010) Worldwide correlations of
#' mechanical properties and green wood density. American Journal of Botany,
#' 97, 1587-1594
#' @keywords datasets
#' @examples
#' 
#' data(Dst)
NULL 



#' Raw CSV file of field recorded values for a stone pine tree
#' 
#' Required data for the assessment of the centre of mass have been recorded in
#' the field for a stone pine (\emph{Pinus pinea} L.). This is an example of
#' csv file that should be fed to \code{\link{treeBiomass}} to assess tree
#' centre of mass.
#'
#' This dataset has been collected for a 17.1 metres tall stone pine whose stem was tilted approx. 20 degrees from the vertical plane (or 80 degrees from the horizontal plane). The stem has been sectioned in two logs (\code{L1} and \code{L2}), and a final branch (\code{C}).
#' 
#' The \code{.csv} file must contain all column headings listed in \code{\link{importFieldData}}, regardless of them being optional (no data in them) or mandatory.
#' 
#' @name stonePine1FieldData
#' @docType data
#' @format \code{
#' "code","azimuth","dBase","dTip","length","tipD","height","tilt","toBePruned","pathToTip"
#' "L1",275,73,41,"10.2","2.5",0,80,"FALSE","TRUE"
#' "L2",275,41,16,"3.9","2.75","10.2",80,"FALSE","TRUE"
#' "B1",190,15,0,"7.95","10.1",,,"FALSE","FALSE"
#' "B2",200,22,0,"7.95","10.4",,,"FALSE","FALSE"
#' "B3",230,15,0,"7.95","10.4",,,"FALSE","FALSE"
#' "B4",200,18,0,"7.95","11.15",,,"FALSE","FALSE"
#' "B5",180,7,0,"7.95","11.3",,,"FALSE","FALSE"
#' "B6",150,6,0,"7.95","11.3",,,"FALSE","FALSE"
#' "B7",340,16,0,"3.95","11.3",,,"FALSE","FALSE"
#' "B8",220,13,0,"7.95","11.8",,,"FALSE","FALSE"
#' "B9",165,19,0,"7.95","11.8",,,"FALSE","FALSE"
#' "B10",280,8,0,"3.95","11.9",,,"FALSE","FALSE"
#' "B11",170,9,0,"7.95","11.9",,,"FALSE","FALSE"
#' "B12",265,8,0,"7.95","12.2",,,"FALSE","FALSE"
#' "B13",75,6,0,"3.95","12.2",,,"FALSE","FALSE"
#' "B14",180,6,0,"7.95","12.2",,,"FALSE","FALSE"
#' "B15",170,6,0,"7.95","12.6",,,"FALSE","FALSE"
#' "B16",120,5,0,"7.95","12.6",,,"FALSE","FALSE"
#' "B17",10,14,0,"3.95",13,,,"FALSE","FALSE"
#' "B18",180,13,0,"7.95",13,,,"FALSE","FALSE"
#' "B19",260,13,0,"7.95","13.2",,,"FALSE","FALSE"
#' "B20",75,6,0,"3.95","13.2",,,"FALSE","FALSE"
#' "B21",75,10,0,"3.95","13.75",,,"FALSE","FALSE"
#' "B22",215,7,0,"7.95","13.75",,,"FALSE","FALSE"
#' "B23",140,7,0,"7.95","13.75",,,"FALSE","FALSE"
#' "C",275,16,0,3,3,"14.1",80,"FALSE","TRUE"
#' }
#' @source Original data collected by the author
#' @keywords datasets
#' @examples
#' library("treecm")
#' csvFileName <- system.file("data", "stonePine1FieldData.csv.gz", package = "treecm")
#' treeData <- importFieldData(
#'   csvFileName, 
#'   650, 
#'   allometryABDC
#' )
#' head(treeData$fieldData)
NULL


#' Field recorded values for a stone pine tree
#' 
#' Required data for the assessment of the centre of mass have been recorded in
#' the field for a stone pine (\emph{Pinus pinea} L.).
#' \code{\link{treeBiomass}} has already been run on the dataset, vectors have
#' yet to be computed.
#' 
#' This dataset includes a list of 4 elements:
#' \itemize{
#' \item{the \code{\link{stonePine1FieldData}} dataset}
#' \item{the density of wood}
#' \item{the allometry function to be used to compute branches biomass}
#' \item{the estimate of branches centre of mass}
#' }
#'
#' @name stonePine1TreeData
#' @docType data
#' @format The format is: 
#' \code{
#'  List of 4
#' $ fieldData   :'data.frame':  26 obs. of  10 variables:
#' ..$ azimuth   : int [1:26] 275 275 190 200 230 200 180 150 340 220 ...
#' ..$ dBase     : int [1:26] 73 41 15 22 15 18 7 6 16 13 ...
#' ..$ dTip      : num [1:26] 41 16 0 0 0 0 0 0 0 0 ...
#' ..$ length    : num [1:26] 10.2 3.9 NA NA NA NA NA NA NA NA ...
#' ..$ tipD      : num [1:26] 2.5 2.75 7.95 7.95 7.95 7.95 7.95 7.95 7.95 3.95 ...
#' ..$ height    : num [1:26] 0 10.2 10.1 10.4 10.4 ...
#' ..$ tilt      : num [1:26] 80 80 0 0 0 0 0 0 0 0 ...
#' ..$ toBePruned: logi [1:26] FALSE FALSE FALSE FALSE FALSE FALSE ...
#' ..$ pathToTip : logi [1:26] TRUE TRUE FALSE FALSE FALSE FALSE ...
#' ..$ biomass   : num [1:26] 1825 193 123 313 123 ...
#' $ density     : num 650
#' $ allometryFUN:function (x, diameter)  
#'   $ branchesCM  : num 1
#' }
#' @source Original data collected by the author
#' @keywords datasets
#' @examples
#' data(stonePine1TreeData)
#' vectors  <- treeVectors(stonePine1TreeData)
#' CM       <- centreOfMass(vectors)
#' summary(CM)
#' # The steps to recreate this dataset:
#' csvFileName <- system.file("data", "stonePine1FieldData.csv.gz", package = "treecm")
#' treeData <- importFieldData(csvFileName, 650, allometryABDC)
#' treeData <- treeBiomass(treeData)
NULL



#' Raw CSV file of field recorded values for a stone pine tree
#' 
#' Required data for the assessment of the centre of mass have been recorded in
#' the field for a stone pine (\emph{Pinus pinea} L.). This is an example of
#' csv file that should be fed to \code{\link{treeBiomass}} to assess tree
#' centre of mass.
#' 
#' This dataset has been collected for a \eqn{\approx 11} metres tall stone pine with a small number of very large branches.
#'
#' The \code{.csv} file must contain all column headings listed in \code{\link{importFieldData}}, regardless of them being optional (no data in them) or mandatory.
#' 
#' @name stonePine2FieldData
#' @docType data
#' @format \code{
#' "code","azimuth","dBase","dTip","length","tipD","height","tilt","toBePruned","pathToTip"
#' "L1",0,67,40,"6.8",0,0,90,,"TRUE"
#' "B1",250,40,,,"7.8","6.8",,,
#' "B2",240,32,,,"8.9","7.8",,,
#' "B3",55,25,,,9,9,,,
#' "B4",260,10,,,"5.5",10,,"TRUE",
#' "B5",80,36,,,"8.2","10.6",,,
#' "B6",255,27,,,"4.5","10.8",,,
#' "B7",0,40,,,"6.5","6.8",70,,"TRUE"
#' }
#' @source Original data collected by the author
#' @keywords datasets
#' @examples
#' library("treecm")
#' csvFileName <- system.file("data", "stonePine2FieldData.csv.gz", package = "treecm")
#' treeData <- importFieldData(csvFileName, 650, allometryABDC)
#' head(treeData$fieldData)
NULL



#' @title Plots a segment
#'
#' @description Plots a segmente given two set of polar coordinates (angle, distance
#' from tree base), may be used to represent buildings close to the tree on the CM plot
#' 
#' @param a0 angle of first set of coordinates
#' @param d0 distance of first set of coordinates
#' @param a1 angle of second set of coordinates
#' @param d1 distance of second set of coordinates
#' @return \code{NULL}
#' @export
plotPolarSegment <- function(a0, d0, a1, d1) {
  xy0 <- toCartesianXY(a0, d0)
  xy1 <- toCartesianXY(a1, d1)
  points(xy0[1], xy0[2])
  points(xy1[1], xy1[2])
  segments(xy0[1], xy0[2], xy1[1], xy1[2])
}

#' @title Imports field data from csv file
#'
#' @description Imports \code{csv} file holding field recorded data returning a list holding field and other key data provided as arguments. 
#'
#' Field measures to be taken on logs and branches include:
#' \itemize{
#'   \item{\bold{Azimuth} (\code{azimuth}): mean angle of orientation of the branch or log measured from the base of the tree (usually with magnetic north as reference, measured clockwise), \emph{mandatory}}
#'   \item{\bold{Diameter at base} (\code{dBase}): diameter at insertion point on the stem, for branches, diameter of the lower section for logs, \emph{mandatory}}
#'   \item{\bold{Diameter at top} (\code{dTip}): always 0 for branches, diameter of the higher section for logs, defaults to 0, \emph{mandatory} only for logs}
#'   \item{\bold{Length} (\code{length}): length of logs or branches, \emph{mandatory} for logs, \emph{mandatory} for branches only if their slenderness coefficient is to be computed}
#'   \item{\bold{Distance} (\code{tipD}): length of branch or log projection on the ground, starting from tree base to tip of branch or log, \emph{mandatory}}
#'   \item{\bold{Height} (\code{height}): height of branch insertion on the stem or height of lower section of the log, to be used to compute \code{z} coordinate of CM, defaults to NA, \emph{mandatory} only for \code{z} determination of CM}
#'   \item{\bold{Tilt} (\code{tilt}): mean branch tilt or log tilt from the horizontal plane (eg a vertical branch is 90 degrees, an horizontal branch is 0 degrees), to be used to compute \code{z} coordinate of CM, defaults to 0, \emph{mandatory} only for \code{z} determination of CM. Note however that the tree tip should be considered as a branch, not a log, in order to account for foliage biomass. In this case the tilt value should be recorded otherwise it would default to 0, \emph{ie} an horizontal branch}
#'   \item{\bold{To be pruned} (\code{toBePruned}): a boolean value, defaults to FALSE, \emph{optional}}
#'   \item{\bold{Path to tip} (\code{pathToTip}): a boolean value TRUE on logs and branches that make part of the ``main stem'' of the tree, defaults to FALSE, \emph{optional}}
#' }
#' 
#' @param fileName Name of csv file holding field data
#' @param dst Fresh density of wood of the tree
#' @param branchesAllometryFUN the function that should compute branch biomass from its diameter
#' @param bCM Estimated position of the centre of mass of branches, a real number from 0.01 (CM at branch base) to 1.00 (CM at branch tip). As a rule of thumb, average live branches, with an average amount of foliage, have CM approx. from \eqn{1/3} to \eqn{2/3} of their length. bCM = 1.0 (default value)
#' @seealso \code{\link{getCoordinatesAndMoment}}
#' @return a list of 4 elements: field data, wood fresh density, allometryFUN function and branches CM
#' @export
importFieldData <- function(fileName, dst, branchesAllometryFUN, bCM = 1) {
  tree <- read.csv(fileName, row.names = 1)
  
  tree <- within(tree, {
    dTip[is.na(dTip)] <- 0
    tilt[is.na(tilt)] <- 0
    toBePruned[is.na(toBePruned)] <- FALSE
    pathToTip[is.na(pathToTip)] <- FALSE
  })
  
  list(
    fieldData    = tree, 
    density      = dst, 
    allometryFUN = branchesAllometryFUN,
    branchesCM   = bCM
  )
  
}

#' @title Stores branches CM in an object
#'
#' @description Stores branches CM in the object provided as an argument. branchesCM has to be in the range 0.01 - 1
#'
#' @note Method \code{\link{treeVectors}} must be invoked to take changes into effect
#'
#' @param object the object of class \code{treeData}
#' @param value the new branch CM
#' @return the updated list
#' @export
setBranchesCM <- function(object, value) {
  if (is.list(object) && value > 0 && value <=1) 
    object$branchesCM <- value
  return(object)
}
